MapboxNavigation/SimulatedRoute.swift

```diff
@@ -9,169 +20,156 @@ class SimulatedLocation: CLLocation {
+                                                    verticalAccuracy: verticalAccuracy,
+                                                    course: wrap(floor(newCoordinate!.direction(to: nextLocation.coordinate)), min: 0, max: 360),
+                                                    speed: speed,
+                                                    timestamp: Date())
+                processedLocations.append(newLocation)
+            }
+        }
+        
+        if processedLocations.last!.coordinate != last!.coordinate {
+            processedLocations.append(last!)
+        }
+        
+        return processedLocations.count > count ? processedLocations.interpolated() : processedLocations
     }
     
-    @objc fileprivate func tick() {
-        guard processedLocations.count > 0 else { return }
-        let location = processedLocations[0] as! SimulatedLocation
-        delegate?.simulation(locationManager, didUpdateLocations: [location.shifted(to: Date())])
+    fileprivate func calculatedTimestamp() -> [SimulatedLocation] {
+        var processedLocations = [SimulatedLocation]()
         
-        if (processedLocations.count > 1) {
-            let nextLocation = processedLocations[1]
-            let delay = nextLocation.timestamp.timeIntervalSince(location.timestamp)
-            perform(#selector(tick), with: nil, afterDelay: delay)
-        } else {
-            // Simulation finished
+        for location in self {
+            if location == first {
+                processedLocations.append(location.shifted(to: Date()))
+            } else {
+                let distance = location.distance(from: processedLocations.last!)
+                let time = distance / location.speed
+                let newLocation = location.shifted(to: processedLocations.last!.timestamp.addingTimeInterval(time))
+                processedLocations.append(newLocation)
+            }
         }
         
-        processedLocations.remove(at: 0)
+        return processedLocations
     }
     
-    func stop() {
-        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(tick), object: nil)
-    }
-}
-
-fileprivate extension Double {
-    func scale(minimumIn: Double, maximumIn: Double, minimumOut: Double, maximumOut: Double) -> Double {
-        return ((maximumOut - minimumOut) * (self - minimumIn) / (maximumIn - minimumIn)) + minimumOut
+    // Calculate speed based on turn penalty.
+    fileprivate func calculatedSpeed() -> [SimulatedLocation] {
+        var processedLocations = [SimulatedLocation]()
+        
+        for location in self {
+            let reversedTurnPenalty = maximumTurnPenalty - location.turnPenalty
+            let speed = reversedTurnPenalty.scale(minimumIn: minimumTurnPenalty, maximumIn: maximumTurnPenalty, minimumOut: minimumSpeed, maximumOut: maximumSpeed)
+            processedLocations.append(location.withSpeed(speed))
```

Comment:
On second thought, maybe we should extend SimulatedLocation with a computed variable `withPenalizedSpeed`:

```swift
var withPenalizedSpeed: SimulatedLocation {
    let reversedTurnPenalty = maximumTurnPenalty - turnPenalty
    let speed = reversedTurnPenalty.scale(minimumIn: minimumTurnPenalty, maximumIn: maximumTurnPenalty, minimumOut: minimumSpeed, maximumOut: maximumSpeed)
    return withSpeed(speed)
}
```

Then the call to `calculatedSpeed()` becomes simply `map(SimulatedLocation.withPenalizedSpeed)`.